src/features/inlay-hints.ts

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2022 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import tsp from 'typescript/lib/protocol.d.js';
+import * as lsp from 'vscode-languageserver';
+import API from '../utils/api.js';
+import type { ConfigurationManager } from '../configuration-manager.js';
+import type { LspDocuments } from '../document.js';
+import type { TspClient } from '../tsp-client.js';
+import type { LspClient } from '../lsp-client.js';
+import { CommandTypes } from '../tsp-command-types.js';
+import { Position } from '../utils/typeConverters.js';
+import { uriToPath } from '../protocol-translation.js';
+
+export class TypeScriptInlayHintsProvider {
+    public static readonly minVersion = API.v440;
+
+    public static async provideInlayHints(
+        uri: lsp.DocumentUri,
+        range: lsp.Range,
+        documents: LspDocuments,
+        tspClient: TspClient,
+        lspClient: LspClient,
+        configurationManager: ConfigurationManager
+    ): Promise<lsp.InlayHint[]> {
+        if (tspClient.apiVersion.lt(TypeScriptInlayHintsProvider.minVersion)) {
+            lspClient.showErrorMessage('Inlay Hints request failed. Requires TypeScript 4.4+.');
+            return [];
+        }
+
+        const file = uriToPath(uri);
+
+        if (!file) {
+            lspClient.showErrorMessage('Inlay Hints request failed. No resource provided.');
+            return [];
+        }
+
+        const document = documents.get(file);
+
+        if (!document) {
+            lspClient.showErrorMessage('Inlay Hints request failed. File not opened in the editor.');
+            return [];
+        }
+
+        if (!areInlayHintsEnabledForFile(configurationManager, file)) {
```

Comment:
is this check necessary?

Reply:
Yes. Otherwise, even with all inlay hint options disabled, the server would still do expensive processing that takes a significant amount of time. The old implementation suffered from that which I've realized when working on this.